uat/testing-features/src/main/java/com/aws/greengrass/steps/BrokerCertificateSteps.java

```diff
@@ -0,0 +1,234 @@
+        X509Certificate certA = certificates.get(certNameA);
+        if (certA == null) {
+            throw new IllegalStateException(String.format("Certificate %s not found.", certNameA));
+        }
+
+        X509Certificate certB = certificates.get(certNameB);
+        if (certB == null) {
+            throw new IllegalStateException(String.format("Certificate %s not found.", certNameB));
+        }
+
+        if (!certA.equals(certB)) {
+            throw new IllegalStateException("Certificates are differ");
+        }
+    }
+
+    /**
+     * Retrive broker's certificate and store it internally for future references.
+     *
+     * @param brokerId the id of broker, before must be discovered or added
+     * @param certName the name (alias) of certificate for future references
+     * @throws InterruptedException when thread has been interrupted
+     * @throws ExecutionException on future's excution exception
+     * @throws TimeoutException when timed out
+     * @throws IllegalStateException on errors
+     */
+    @When("I retrieve the certificate of broker {string} and store as {string}")
+    public void retrieveServerCertificate(String brokerId, String certName)
+                throws ExecutionException, TimeoutException, InterruptedException {
+
+        // get address information about broker
+        final List<MqttBrokers.ConnectivityInfo> bc = mqttBrokers.getConnectivityInfo(brokerId);
+        if (CollectionUtils.isNullOrEmpty(bc)) {
+            throw new IllegalStateException("There is no address information about broker, "
+                                        + "probably discovery step missing in scenario");
+        }
+
+        Exception lastException = null;
+        for (MqttBrokers.ConnectivityInfo broker : bc) {
+            String host = broker.getHost();
+            Integer port = broker.getPort();
+            URI serverUri = URI.create(String.format("https://%s:%d", host, port));
+
+            try {
+                CertificateFutures futures = getServerCerts(serverUri);
+                X509Certificate[] serverCerts = futures.getCertificate().get(GET_CERTIFICATE_TIMEOUT_SEC,
+                                                                            TimeUnit.SECONDS);
+                if (serverCerts == null) {
+                    throw new IllegalStateException("Certificate array is null");
+                }
+
+                if (serverCerts.length <= 0) {
+                    throw new IllegalStateException("Certificate array is empty");
+                }
+
+                X509Certificate cert = serverCerts[0];
+                certificates.put(certName, cert);
+                log.info("Saved broker's '{}' certificate '{}'", brokerId, cert);
+                return;
+            } catch (IllegalStateException | ExecutionException | TimeoutException ex) {
+                lastException = ex;
+            }
+        }
+        if (lastException == null) {
```

Comment:
could you please add a comment why the null exception is not logically correct.

Reply:
Added in e1cede5de9a5ae4e77b39a8fb35f97e0d401406f